### PR TITLE
Fix Issue 4946 - Not good error message with wrongly positioned 'const'

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/ddmd/parse.d [hunk 1]
```diff
@@ -4767,6 +4767,13 @@ final class Parser(AST) : Lexer
             else if (!f.frequire && !f.fensure) // allow these even with no body
             {
                 error("semicolon expected following function declaration");
+                TOK t = token.value;
+                import core.stdc.stdio;
```
Comment (wilzbach) [inline src/ddmd/parse.d]: Remove

File: src/ddmd/parse.d [hunk 2]
```diff
@@ -4767,6 +4767,13 @@ final class Parser(AST) : Lexer
...
+                if (t == TOKconst || t == TOKimmutable || t == TOKinout || t == TOKreturn ||
+                    t == TOKshared || t == TOKnothrow || t == TOKpure)
+                    error("'%s' cannot be placed after template constraint", token.toChars);
```
Comment (wilzbach) [inline src/ddmd/parse.d]: after __a__ template constraint?

File: src/ddmd/parse.d [hunk 3]
```diff
@@ -4767,6 +4767,13 @@ final class Parser(AST) : Lexer
             else if (!f.frequire && !f.fensure) // allow these even with no body
             {
                 error("semicolon expected following function declaration");
```
Comment (wilzbach) [inline src/ddmd/parse.d]: How about moving this to the `else` - i.e. as default message?

**Review by wilzbach — APPROVED** [review]
Looks reasonable to me.

File: src/ddmd/parse.d (hunk 1 above)
Comment (RazvanN7) [reply src/ddmd/parse.d]: oh yeah, thanks, forgot about it.

File: src/ddmd/parse.d (hunk 3 above)
Comment (RazvanN7) [reply src/ddmd/parse.d]: fair enough

File: src/ddmd/parse.d [hunk 4]
```diff
@@ -4766,7 +4766,14 @@ final class Parser(AST) : Lexer
             }
             else if (!f.frequire && !f.fensure) // allow these even with no body
             {
-                error("semicolon expected following function declaration");
+                TOK t = token.value;
+                if (t == TOKconst || t == TOKimmutable || t == TOKinout || t == TOKreturn ||
+                    t == TOKshared || t == TOKnothrow || t == TOKpure)
+                    error("'%s' cannot be placed after a template constraint", token.toChars);
+                else if (t == TOKat)
+                    error("annotations cannot be placed after a template constraint");
```
Comment (BBasile) [inline src/ddmd/parse.d]: rather _attributes_ no ? _annotation_  isn't an official name.

Comment (RazvanN7) [reply src/ddmd/parse.d]: yeah, it does sound better. thanks!

**Review by andralex — APPROVED** [review]

**Pull request merged**